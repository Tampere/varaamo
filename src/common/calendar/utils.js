import moment from 'moment';
import get from 'lodash/get';

/**
 * getDefaultSelectedTimeRange()
 * Get default selected time range from reservation data.
 * @returns {Object} selected time range
 */
export const getDefaultSelectedTimeRange = (reservation) => {
  return reservation ? {
    start: reservation.begin,
    end: reservation.end,
  } : null;
};


/**
 * isTimeRangeOverMaxPeriod();
 * Reservation cannot be longer than the resources max period if max period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than max_period or not
 */
export const isTimeRangeOverMaxPeriod = (resource, start, end, isStaff = false) => {
  const period = get(resource, 'max_period', null);

  if (!isStaff && period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') > durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * isTimeRangeUnderMinPeriod();
 * Reservation must longer than the resources min period if min period is set.
 *
 * @param {Object} resource Resource data
 * @param {Date} start Start time
 * @param {Date} end End time
 * @param {boolean} isStaff Bypass this check if user is staff member
 * @returns {boolean} Selected time range is larger than min_period or not
 */
export const isTimeRangeUnderMinPeriod = (resource, start, end) => {
  const period = get(resource, 'min_period', null);

  if (period) {
    const startMoment = moment(start);
    const endMoment = moment(end);

    const periodDuration = moment.duration(period);
    const durationInMinutes = periodDuration.hours() * 60 + periodDuration.minutes();

    if (endMoment.diff(startMoment, 'minutes') < durationInMinutes) {
      return true;
    }
  }

  return false;
};

/**
 * addPeriodToTime()
 * Add period to input time
 * For example: period 00:30:00, time: moment(12:00:00) -> moment(12:30:00)
 *
 * @param {Moment} time
 * @param {string} period
 * @returns {Moment}
 */

const addPeriodToTime = (time, period) => {
  if (period) {
    const periodInMinute = moment.duration(period).as('minutes');
    return time.add(periodInMinute, 'minutes');
  }

  // Return original if no period
  return time;
};

/**
 * getMinPeriodTimeRange
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMinPeriodTimeRange = (resource, start, end) => {
  const minPeriod = get(resource, 'min_period', null);

  if (!minPeriod) {
    return {
      start, end,
    };
  }

  const minPeriodEnd = addPeriodToTime(moment(start), minPeriod);

  if (minPeriodEnd.diff(moment(end)) > 0) {
    // time range smaller than min_period, return min_period
    return {
      start,
      end: minPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};

/**
 * getMaxPeriodEndTime
 * Populate the end time with resource min_period
 * For example: start time 1pm, min_period: 1:00:00 ->
 * end time should be 1pm + 1hour = 2pm
 *
 * @param {Object} resource
 * @param {Date} start
 * @param {Date} end
 * @returns {Date} return new end date if time range < min_period given, otherwise return original end
 */

export const getMaxPeriodTimeRange = (resource, start, end, isStaff) => {
  const maxPeriod = get(resource, 'max_period', null);

  if (!maxPeriod || isStaff) {
    return {
      start, end,
    };
  }

  const maxPeriodEnd = addPeriodToTime(moment(start), maxPeriod);

  if (moment(end).diff(maxPeriodEnd) > 0) {
    // time range is over max_period not allowed, return max_period
    return {
      start,
      end: maxPeriodEnd.toDate(),
    };
  }

  return {
    start,
    end,
  };
};
