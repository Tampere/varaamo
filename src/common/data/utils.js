import get from 'lodash/get';
import values from 'lodash/values';
import pickBy from 'lodash/pickBy';
import camelCase from 'lodash/camelCase';

export const getLocalizedFieldValue = (field, locale, fallback = false) => {
  const localeValue = get(field, locale, null);

  if (localeValue || !fallback) {
    return localeValue;
  }

  return values(field).find(fallbackValue => !!fallbackValue);
};

/**
 * Trim empty values from data object
 *
 * @param {Object} data
 * @returns {Object} Trimmed empty values object
 */

export const parseData = (data) => {
  return pickBy(data, value => value || value === 0);
};

export const toCamelCase = (obj) => {
  if (!obj) return {};

  const camelCasedObj = {};

  Object.keys(obj).forEach((key) => {
    const camelCasedKey = camelCase(key);
    camelCasedObj[camelCasedKey] = obj[key];
  });

  return camelCasedObj;
};
