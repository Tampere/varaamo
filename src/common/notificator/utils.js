import moment from 'moment';

/**
 * createSelectOptions()
 * Create option arrays for selects.
 * @returns {Object} with select option arrays
 */
export const createSelectOptions = () => {
  const targetOptions = [
    { value: 'staff', label: 'Staff' },
    { value: 'user', label: 'User' },
    { value: 'all', label: 'All' },
  ];
  const urgencyOptions = [
    { value: 'common', label: 'Common' },
    { value: 'warning', label: 'Warning' },
    { value: 'danger', label: 'Danger' },
  ];
  const activeOptions = [
    { value: true, label: 'True' },
    { value: false, label: 'False' },
  ];

  return {
    targetOptions,
    urgencyOptions,
    activeOptions,
  };
};


/**
 * formatValuesForUse()
 * Format notification values to usable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForUse = (notification) => {
  const { targetOptions, urgencyOptions, activeOptions } = createSelectOptions();
  const formattedNotification = { ...notification };
  formattedNotification.target = targetOptions.find(option => option.value === notification.target);
  formattedNotification.urgency = urgencyOptions.find(option => option.value === notification.urgency);
  formattedNotification.active = activeOptions.find(option => option.value === notification.active);
  formattedNotification.until = moment(notification.until).toDate();
  return formattedNotification;
};

/**
 * formatValuesForDatabase()
 * Format notification values to savable form.
 * @returns {Object} with formatted values
 */
export const formatValuesForDatabase = (notification) => {
  const formattedNotification = { ...notification };
  formattedNotification.target = notification.target.value;
  formattedNotification.urgency = notification.urgency.value;
  formattedNotification.active = notification.active.value;
  formattedNotification.until = moment(notification.until).format('YYYYMMDDTHHmmss');
  return formattedNotification;
};
