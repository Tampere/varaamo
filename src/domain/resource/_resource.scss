@import './card/resourceCard';
@import './card/slider/resourceCardSlider';
@import './map/resourceMap';
@import './reservationCalendar/resourceReservationCalendar';
