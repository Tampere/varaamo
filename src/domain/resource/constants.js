export const resourceProductTypes = Object.freeze({
  RENT: 'rent',
  EXTRA: 'extra',
});

export const resourcePriceTypes = Object.freeze({
  HOURLY: 'hourly',
  DAILY: 'daily',
  WEEKLY: 'weekly',
  FIXED: 'fixed',
});
