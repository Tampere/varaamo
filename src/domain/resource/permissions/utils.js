import difference from 'lodash/difference';
import get from 'lodash/get';

import {
  resourceRoles, resourcePermissionTypes, resourcePermissionsByRole, UI_UNIT_STAFF_ROLES,
} from './constants';

const permissionOptions = Object.values(resourcePermissionTypes);
const allValuesInArray = (subset, superset) => difference(subset, superset).length === 0;

export function hasPermissionForResource(resourceRole, requiredPermissions) {
  // undefined roles don't receive any permissions implicitly
  if (!resourceRole) {
    return false;
  }

  if (!Object.values(resourceRoles).includes(resourceRole)) {
    throw Error(`resourceRole needs to be one of: ${Object.values(resourceRoles).join(', ')}.`);
  }

  if (!Array.isArray(requiredPermissions) && typeof requiredPermissions !== 'string') {
    throw Error('requiredPermissions has to be either a string or an array.');
  }

  if (
    Array.isArray(requiredPermissions)
    && !allValuesInArray(requiredPermissions, permissionOptions)
  ) {
    const unknownPermissions = difference(requiredPermissions, permissionOptions);

    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${unknownPermissions.join(', ')}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  if (typeof requiredPermissions === 'string' && !permissionOptions.includes(requiredPermissions)) {
    // eslint-disable-next-line no-console
    console.warn([
      `Permission asked with unknown permission: ${requiredPermissions}.`,
      `Permission needs to be on of: ${permissionOptions.join(', ')}.`,
    ].join('\n'));
  }

  const rolePermissions = resourcePermissionsByRole[resourceRole];

  if (Array.isArray(requiredPermissions)) {
    return allValuesInArray(requiredPermissions, rolePermissions);
  }

  return rolePermissions.includes(requiredPermissions);
}

const getUserPermissions = resource => (
  get(resource, 'user_permissions', false)
  || get(resource, 'userPermissions', undefined)
);

// Returns the role with the most permissions or null.
export function getUnitRoleFromResource(resource) {
  const userPermissions = getUserPermissions(resource);

  if (!userPermissions) {
    return null;
  }

  if (userPermissions.is_admin || userPermissions.isAdmin) {
    return resourceRoles.UNIT_ADMINISTRATOR;
  }

  if (userPermissions.is_manager || userPermissions.isManager) {
    return resourceRoles.UNIT_MANAGER;
  }

  if (userPermissions.is_viewer || userPermissions.isViewer) {
    return resourceRoles.UNIT_VIEWER;
  }

  return null;
}

export function getIsUnitStaff(unitRole) {
  return UI_UNIT_STAFF_ROLES.includes(unitRole);
}
