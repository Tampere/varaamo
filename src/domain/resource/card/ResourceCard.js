import React from 'react';
import { connect } from 'react-redux';
import { createStructuredSelector } from 'reselect';
import { Link } from 'react-router-dom';
import { injectIntl, intlShape } from 'react-intl';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import flowRight from 'lodash/flowRight';
import iconHome from 'hel-icons/dist/shapes/home.svg';
import iconMapMarker from 'hel-icons/dist/shapes/map-marker.svg';
import iconTicket from 'hel-icons/dist/shapes/ticket.svg';
import iconUser from 'hel-icons/dist/shapes/user-o.svg';
import iconHeart from 'hel-icons/dist/shapes/heart-o.svg';

import injectT from '../../../../app/i18n/injectT';
import iconMap from '../../../../app/assets/icons/map.svg';
import iconHeartFilled from '../../../../app/assets/icons/heart-filled.svg';
import * as dataUtils from '../../../common/data/utils';
import * as urlUtils from '../../../common/url/utils';
import * as searchUtils from '../../search/utils';
import * as resourceUtils from '../utils';
import BackgroundImage from '../../../../app/shared/background-image/BackgroundImage';
import { getMainImage } from '../../../../app/utils/imageUtils';
import ResourceAvailability from '../availability/ResourceAvailability';
import UnpublishedLabel from '../../../../app/shared/label/unpublished/UnpublishedLabel';
import ResourceCardInfoCell from './ResourceCardInfoCell';
import { isLoggedInSelector } from '../../../../app/state/selectors/authSelectors';


class ResourceCard extends React.Component {
  static propTypes = {
    date: PropTypes.string.isRequired,
    resource: PropTypes.object.isRequired,
    unit: PropTypes.object.isRequired,
    t: PropTypes.func.isRequired,
    history: PropTypes.object,
    intl: intlShape.isRequired,
    isStacked: PropTypes.bool,
    isLoggedIn: PropTypes.bool,
    onFavoriteClick: PropTypes.func.isRequired,
    onFilterClick: PropTypes.func.isRequired,
  };

  getResourcePageLink = () => {
    const { date, resource } = this.props;

    return {
      pathname: resourceUtils.getResourcePageLink(resource),
      search: date ? urlUtils.getSearch({ date }) : '',
      state: { fromSearchResults: true },
    };
  };

  onSearchClick = (filters) => {
    const { history } = this.props;
    history.push(searchUtils.getSearchPageLink(filters));
  };

  render() {
    const {
      date,
      isLoggedIn,
      isStacked,
      resource,
      onFavoriteClick,
      onFilterClick,
      unit,
      intl,
      t,
    } = this.props;
    const locale = intl.locale;
    const typeName = resource.type ? dataUtils.getLocalizedFieldValue(resource.type.name, locale) : '';

    return (
      <div
        className={classNames('app-resourceCard', {
          'app-resourceCard__stacked': isStacked,
        })}
      >
        <Link className="app-resourceCard__image-link" onClick={this.handleLinkClick} to={this.getResourcePageLink()}>
          <BackgroundImage height={420} image={getMainImage(resource.images)} width={700} />
        </Link>
        <div className="app-resourceCard__content">
          <div className="app-resourceCard__unit-name">
            <a
              className="app-resourceCard__unit-name-link"
              onClick={() => this.onSearchClick({ unit: resource.unit })}
              role="button"
              tabIndex="-1"
            >
              <span>{dataUtils.getLocalizedFieldValue(unit.name, locale)}</span>
            </a>
            <div className="app-resourceCard__unit-name-labels">
              <ResourceAvailability date={date} resource={resource} />
              {!resource.public && <UnpublishedLabel />}
            </div>
          </div>
          <Link onClick={this.handleLinkClick} to={this.getResourcePageLink()}>
            <h4>{dataUtils.getLocalizedFieldValue(resource.name, locale, true)}</h4>
          </Link>
          <div className="app-resourceCard__description">
            {dataUtils.getLocalizedFieldValue(resource.description, locale, true)}
          </div>
        </div>
        <div className="app-resourceCard__info">
          {resource.type && (
            <ResourceCardInfoCell
              icon={iconHome}
              onClick={() => onFilterClick('search', typeName)}
              text={typeName}
            />
          )}
          <ResourceCardInfoCell
            icon={iconUser}
            onClick={
              resource.people_capacity
                ? () => onFilterClick('people', searchUtils.getClosestPeopleCapacityValue(resource.people_capacity))
                : null
            }
            text={
              resource.people_capacity
                ? t('ResourceCard.peopleCapacity', { people: resource.people_capacity })
                : '-'
            }
          />
          <ResourceCardInfoCell
            icon={iconTicket}
            onClick={resourceUtils.isFree(resource) ? () => onFilterClick('freeOfCharge', true) : null}
            text={resourceUtils.getPriceFromSnakeCaseResource(resource, t)}
          />
          <ResourceCardInfoCell
            icon={iconMap}
            text={resourceUtils.getUnitAddress(unit, locale)}
          />
          {resource.distance && (
            <ResourceCardInfoCell
              icon={iconMapMarker}
              onClick={() => null}
              text={resourceUtils.getResourceDistance(resource)}
            />
          )}
          {isLoggedIn && (
            <ResourceCardInfoCell
              alt={typeName}
              icon={resource.is_favorite ? iconHeartFilled : iconHeart}
              onClick={() => onFavoriteClick(resource)}
            />
          )}
        </div>
      </div>
    );
  }
}

const UnconnectedResourceCard = injectT(ResourceCard);
export { UnconnectedResourceCard };

const selector = createStructuredSelector({
  isLoggedIn: isLoggedInSelector,
});

export default flowRight([
  injectIntl,
  connect(selector),
])(UnconnectedResourceCard);
