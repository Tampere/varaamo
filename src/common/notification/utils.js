import { NotificationManager } from 'react-notifications';

import { NOTIFICATION_TYPE } from './constants';
/**
 * createNotification()
 * Creating notification popups
 *
 * @param {string} type
 * @param {string} [message='']
 * @param {string} [title='']
 * @param {number} [fadeTime=2000]
 * @param {function} callback
 * @returns {void}
 */
export const createNotification = (type, message = '', title = '', fadeTime = 2000, callback) => {
  switch (type) {
    case NOTIFICATION_TYPE.INFO:
      NotificationManager.info(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.SUCCESS:
      NotificationManager.success(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.WARNING:
      NotificationManager.warning(message, title, fadeTime, callback);
      break;
    case NOTIFICATION_TYPE.ERROR:
      NotificationManager.error(message, title, fadeTime, callback);
      break;
    default:
      NotificationManager.info(message, title, fadeTime, callback);
  }
};
