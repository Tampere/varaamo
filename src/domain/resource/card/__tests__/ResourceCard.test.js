import React from 'react';
import toJSON from 'enzyme-to-json';

import { shallowWithIntl } from '../../../../../app/utils/testUtils';
import { UnconnectedResourceCard } from '../ResourceCard';
import unit from '../../../../common/data/fixtures/unit';
import resource from '../../../../common/data/fixtures/resource';

describe('ResourceCard', () => {
  test('renders correctly', () => {
    const props = {
      date: '30-07-2019',
      key: 'foo',
      onFavoriteClick: jest.fn(),
      onFilterClick: jest.fn(),
      resource: resource.build(),
      unit: unit.build(),
      t: () => {},
    };
    const wrapper = shallowWithIntl(
      <UnconnectedResourceCard {...props} />,
    );

    expect(toJSON(wrapper)).toMatchSnapshot();
  });
});
