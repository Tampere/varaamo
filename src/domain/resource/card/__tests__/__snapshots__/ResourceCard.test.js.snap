// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ResourceCard renders correctly 1`] = `
<div
  className="app-resourceCard"
>
  <Link
    className="app-resourceCard__image-link"
    replace={false}
    to={
      Object {
        "pathname": "/resources/1",
        "search": "date=30-07-2019",
        "state": Object {
          "fromSearchResults": true,
        },
      }
    }
  >
    <BackgroundImage
      height={420}
      image={Object {}}
      width={700}
    />
  </Link>
  <div
    className="app-resourceCard__content"
  >
    <div
      className="app-resourceCard__unit-name"
    >
      <a
        className="app-resourceCard__unit-name-link"
        onClick={[Function]}
        role="button"
        tabIndex="-1"
      >
        <span>
          Unit (en) - 1
        </span>
      </a>
      <div
        className="app-resourceCard__unit-name-labels"
      >
        <InjectT(ResourceAvailability)
          date="30-07-2019"
          resource={
            Object {
              "available_hours": Array [],
              "equipment": Array [],
              "id": 1,
              "is_favorite": false,
              "name": Object {
                "en": "Resource (en) - 1",
                "fi": "Resource (fi) - 1",
                "sv": "Resource (sv) - 1",
              },
              "need_manual_confirmation": false,
              "opening_hours": Array [],
              "products": Array [],
              "public": true,
              "required_reservation_extra_fields": Array [],
              "reservable": true,
              "reservable_after": null,
              "slot_size": "00:30:00",
              "supported_reservation_extra_fields": Array [],
              "unit": 1,
              "user_permissions": Object {
                "canMakeReservations": true,
                "isAdmin": false,
              },
            }
          }
        />
      </div>
    </div>
    <Link
      replace={false}
      to={
        Object {
          "pathname": "/resources/1",
          "search": "date=30-07-2019",
          "state": Object {
            "fromSearchResults": true,
          },
        }
      }
    >
      <h4>
        Resource (en) - 1
      </h4>
    </Link>
    <div
      className="app-resourceCard__description"
    />
  </div>
  <div
    className="app-resourceCard__info"
  >
    <ResourceCardInfoCell
      icon="test-file-stub"
      onClick={null}
      text="-"
    />
    <ResourceCardInfoCell
      icon="test-file-stub"
      onClick={[Function]}
      text="ResourceIcons.free"
    />
    <ResourceCardInfoCell
      icon="test-file-stub"
      text="null"
    />
  </div>
</div>
`;
